### PR TITLE
Feature/searching

### PR DESCRIPTION
**Pull request by OlegMR0**

### CODE REVIEW

File: src/main/java/com/example/bookstore/repository/book/PriceSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    public String getKey() {
+        return FIELD_SPECIFICATION;
+    }
+
+    public Specification<Book> getSpecification(List<String> params) {
+        return new Specification<Book>() {
+            @Override
+            public Predicate toPredicate(Root<Book> root, CriteriaQuery<?> query,
+                                         CriteriaBuilder criteriaBuilder) {
+                return root.get(FIELD_SPECIFICATION).in(params);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/repository/book/PriceSpecificationProvider.java]: How is it possible to have a list of price values? You should have min/max price boundaries

File: src/main/java/com/example/bookstore/repository/book/TitleSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        return FIELD_SPECIFICATION;
+    }
+
+    @Override
+    public Specification<Book> getSpecification(List<String> params) {
+        return new Specification<Book>() {
+            @Override
+            public Predicate toPredicate(Root<Book> root, CriteriaQuery<?> query,
+                                         CriteriaBuilder criteriaBuilder) {
+                return root.get(FIELD_SPECIFICATION).in(params);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/repository/book/TitleSpecificationProvider.java]: Len's implement 'contains' logic for Title

File: src/main/java/com/example/bookstore/service/BookServiceImpl.java [hunk 3]
```diff
@@ -33,6 +37,30 @@ public List<BookDto> findAll() {
...
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        bookRepository.deleteById(id);
+    }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto book) {
+        if (bookRepository.findById(id).isPresent()) {
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/BookServiceImpl.java]: use `existsById(id)`

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bookstore/repository/BookSpecificationBuilder.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        specification = specification.and(parseListToSpecification(searchParameters.getAuthor(),
+                "author"));
+        specification = specification.and(parseListToSpecification(searchParameters.getIsbn(),
+                "isbn"));
+        List<String> priceBounds = Arrays.asList(searchParameters.getPriceFrom(),
+                searchParameters.getPriceTo());
+        specification = specification.and(parseListToSpecification(priceBounds, "price"));
+        specification = specification.and(parseListToSpecification(searchParameters.getTitle(),
+                "title"));
+        return specification;
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/repository/BookSpecificationBuilder.java:30]: Use constants

File: src/main/java/com/example/bookstore/repository/BookSpecificationProviderManager.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import com.example.bookstore.model.Book;
+import java.util.List;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@AllArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private static final String ERROR_MESSAGE = "Can't find correct "
+            + "specification provider for this key ";
```
Comment (Sanderfuck) [inline src/main/java/com/example/bookstore/repository/BookSpecificationProviderManager.java:12]: Try to use String format

**Review by Sanderfuck — COMMENTED** [review]
Check comments

**Comment by OlegMR0** [comment]
#6 Pull request

**Pull request merged**